File: Packages/Sources/ClarcCore/Models/TodoItem.swift
import Foundation

// MARK: - Todo Item

public struct TodoItem: Identifiable, Sendable, Equatable, Codable {
    public enum Status: String, Sendable, Codable {
        case pending
        case inProgress = "in_progress"
        case completed
    }

    public let id: Int
    public let content: String
    public let activeForm: String
    public let status: Status

    public init(id: Int, content: String, activeForm: String, status: Status) {
        self.id = id
        self.content = content
        self.activeForm = activeForm
        self.status = status
    }
}

// MARK: - Todo Extractor

public enum TodoExtractor {
    /// Find the most recent `TodoWrite` tool call across the message list and
    /// parse its `todos` input into typed ``TodoItem`` values.
    ///
    /// Returns `nil` if no TodoWrite has been emitted yet, or if its `input`
    /// has not finished streaming. Returns an empty array if the call exists
    /// but contains no todos.
    public static func latest(in messages: [ChatMessage]) -> [TodoItem]? {
        for message in messages.reversed() {
            for block in message.blocks.reversed() {
                guard let toolCall = block.toolCall,
                      toolCall.name.lowercased() == "todowrite" else { continue }
                return parse(input: toolCall.input)
            }
        }
        return nil
    }

    /// Parse a raw `TodoWrite` `input` dictionary into typed ``TodoItem`` values.
    public static func parse(input: [String: JSONValue]) -> [TodoItem] {
        guard let array = input["todos"]?.arrayValue else { return [] }
        return array.enumerated().map { index, value in
            let content = value["content"]?.stringValue ?? ""
            let activeForm = value["activeForm"]?.stringValue ?? content
            let rawStatus = value["status"]?.stringValue ?? ""
            let status = TodoItem.Status(rawValue: rawStatus) ?? .pending
            return TodoItem(id: index, content: content, activeForm: activeForm, status: status)
        }
    }
}

File: RxCodeAndroid/app/src/main/java/app/rxlab/rxcode/proto/TodoItem.kt
package app.rxlab.rxcode.proto

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

@Serializable
data class TodoItem(
    val id: Int,
    val content: String,
    val activeForm: String,
    val status: Status,
) {
    @Serializable
    enum class Status {
        @SerialName("pending") PENDING,
        @SerialName("in_progress") IN_PROGRESS,
        @SerialName("completed") COMPLETED,
    }
}

/**
 * Locates the most recent `TodoWrite` tool call across the message list and
 * parses its `todos` array into typed [TodoItem]s. Mirrors the iOS
 * `TodoExtractor.latest` so the navigation-title progress ring and the
 * todo/summary sheet reflect the same data the desktop sees.
 */
object TodoExtractor {
    fun latest(messages: List<ChatMessage>): List<TodoItem>? {
        for (message in messages.asReversed()) {
            for (block in message.blocks.asReversed()) {
                val toolCall = block.toolCall ?: continue
                if (toolCall.name.equals("TodoWrite", ignoreCase = true)) {
                    return parse(toolCall.input)
                }
            }
        }
        return null
    }

    fun parse(input: JsonObject): List<TodoItem> {
        val array = input["todos"] as? JsonArray ?: return emptyList()
        return array.mapIndexedNotNull { index, element ->
            val obj = element as? JsonObject ?: return@mapIndexedNotNull null
            val content = obj.string("content").orEmpty()
            val activeForm = obj.string("activeForm").takeIf { !it.isNullOrEmpty() } ?: content
            val status = when (obj.string("status")) {
                "in_progress" -> TodoItem.Status.IN_PROGRESS
                "completed" -> TodoItem.Status.COMPLETED
                else -> TodoItem.Status.PENDING
            }
            TodoItem(id = index, content = content, activeForm = activeForm, status = status)
        }
    }

    private fun JsonObject.string(key: String): String? =
        (this[key] as? JsonPrimitive)?.contentOrNull
}
